polaris-react/src/components/Scrollable/Scrollable.tsx

```diff
@@ -40,218 +38,93 @@ export interface ScrollableProps extends React.HTMLProps<HTMLDivElement> {
-  }
-
-  componentDidUpdate() {
-    const {scrollPosition} = this.state;
-    if (scrollPosition && this.scrollArea && scrollPosition > 0) {
-      this.scrollArea.scrollTop = scrollPosition;
-    }
-  }
-
-  render() {
-    const {topShadow, bottomShadow, canScroll} = this.state;
-    const {
-      children,
-      className,
-      horizontal = true,
-      vertical = true,
-      shadow,
-      hint,
-      focusable,
-      onScrolledToBottom,
-      ...rest
-    } = this.props;
-
-    const finalClassName = classNames(
-      className,
-      styles.Scrollable,
-      vertical && styles.vertical,
-      horizontal && styles.horizontal,
-      topShadow && styles.hasTopShadow,
-      bottomShadow && styles.hasBottomShadow,
-      vertical && canScroll && styles.verticalHasScrolling,
-    );
+  useEffect(() => {
+    const currentScrollArea = scrollArea.current;
 
-    return (
-      <ScrollableContext.Provider value={this.scrollToPosition}>
-        <StickyManagerContext.Provider value={this.stickyManager}>
-          <div
-            className={finalClassName}
-            {...scrollable.props}
-            {...rest}
-            ref={this.setScrollArea}
-            // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
-            tabIndex={focusable ? 0 : undefined}
-          >
-            {children}
-          </div>
-        </StickyManagerContext.Provider>
-      </ScrollableContext.Provider>
-    );
-  }
-
-  private setScrollArea = (scrollArea: HTMLElement | null) => {
-    this.scrollArea = scrollArea;
-  };
-
-  private handleScroll = () => {
-    const {scrollArea} = this;
-    const {scrollPosition} = this.state;
-    const {shadow, onScrolledToBottom} = this.props;
-    if (scrollArea == null) {
+    if (!currentScrollArea) {
       return;
     }
-    const {scrollTop, clientHeight, scrollHeight} = scrollArea;
-    const shouldBottomShadow = Boolean(
-      shadow && !(scrollTop + clientHeight >= scrollHeight),
-    );
-    const shouldTopShadow = Boolean(
-      shadow && scrollTop > 0 && scrollPosition > 0,
-    );
 
-    const canScroll = scrollHeight > clientHeight;
-    const hasScrolledToBottom =
-      scrollHeight - scrollTop <= clientHeight + LOW_RES_BUFFER;
+    const handleScroll = () => {
+      const {scrollTop, clientHeight, scrollHeight} = currentScrollArea;
+      const isBelowTopOfScroll = Boolean(scrollTop > 0);
+      const isAtBottomOfScroll = Boolean(
+        scrollTop + clientHeight >= scrollHeight - LOW_RES_BUFFER,
+      );
 
-    if (canScroll && hasScrolledToBottom && onScrolledToBottom) {
-      onScrolledToBottom();
-    }
+      setTopShadow(isBelowTopOfScroll);
+      setBottomShadow(!isAtBottomOfScroll);
 
-    this.setState({
-      topShadow: shouldTopShadow,
-      bottomShadow: shouldBottomShadow,
-      scrollPosition: scrollTop,
-      canScroll,
-    });
-  };
+      if (isAtBottomOfScroll && onScrolledToBottom) {
+        onScrolledToBottom();
```

Comment:
I noticed this too and checked with the old implementation. I suspect we'd want to keep it this way as changing could break some consumer. But agree this would ideally be called only once (maybe with retry capability).

Reply:
Following up for posterity, I believe this happens because we receive multiple scroll events within the "buffer zone" for the callback. Original PR that introduced that for reference https://github.com/Shopify/polaris/pull/4218.